File: scala/src/main/scala/folio/codinginterview/presentation/OrderController.scala
package folio.codinginterview.presentation

import folio.codinginterview.application.usecase.order.AdditionalBuyOrderUsecase
import folio.codinginterview.application.usecase.order.AdditionalBuyOrderUsecaseException
import folio.codinginterview.application.usecase.order.AdditionalBuyOrderUsecaseInput
import folio.codinginterview.application.usecase.order.NewOrderUsecase
import folio.codinginterview.application.usecase.order.NewOrderUsecaseException
import folio.codinginterview.application.usecase.order.NewOrderUsecaseInput
import folio.codinginterview.presentation.PresentationException.BadRequestException
import scala.concurrent.ExecutionContext
import scala.concurrent.Future

object OrderController {
  final case class NewOrderRequest(userId: String, amount: String)
  final case class AdditionalOrderRequest(userId: String, amount: String)
}

final class OrderController(
    newOrderUsecase: NewOrderUsecase,
    additionalBuyOrderUsecase: AdditionalBuyOrderUsecase
)(using ec: ExecutionContext)
    extends PresentationPreparation {
  import OrderController.*

  def newOrder(req: NewOrderRequest): Future[Unit] =
    for {
      uid <- parseUserId(req.userId)
      amt <- parseAmount(req.amount)
      _ <- newOrderUsecase.run(NewOrderUsecaseInput(uid, amt)).recoverWith {
        case NewOrderUsecaseException.UserAlreadyExists =>
          Future.failed(BadRequestException("user already has account"))
        case NewOrderUsecaseException.AmountTooSmall =>
          Future.failed(BadRequestException("amount is too small"))
      }
    } yield ()

  def additionalOrder(req: AdditionalOrderRequest): Future[Unit] =
    for {
      uid <- parseUserId(req.userId)
      amt <- parseAmount(req.amount)
      _ <- additionalBuyOrderUsecase.run(AdditionalBuyOrderUsecaseInput(uid, amt)).recoverWith {
        case AdditionalBuyOrderUsecaseException.UserNotFound =>
          Future.failed(BadRequestException("user has no live account"))
        case AdditionalBuyOrderUsecaseException.AmountTooSmall =>
          Future.failed(BadRequestException("amount is too small"))
      }
    } yield ()
}

File: php/src/CodingInterview/Presentation/OrderController.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Presentation;

use Folio\CodingInterview\Application\Usecase\Order\AdditionalBuyOrderAmountTooSmallException;
use Folio\CodingInterview\Application\Usecase\Order\AdditionalBuyOrderUserNotFoundException;
use Folio\CodingInterview\Application\Usecase\Order\AdditionalBuyOrderUsecase;
use Folio\CodingInterview\Application\Usecase\Order\AdditionalBuyOrderUsecaseInput;
use Folio\CodingInterview\Application\Usecase\Order\NewOrderAmountTooSmallException;
use Folio\CodingInterview\Application\Usecase\Order\NewOrderUserAlreadyExistsException;
use Folio\CodingInterview\Application\Usecase\Order\NewOrderUsecase;
use Folio\CodingInterview\Application\Usecase\Order\NewOrderUsecaseInput;

final class NewOrderRequest
{
    public function __construct(
        public readonly string $userId,
        public readonly string $amount,
    ) {}
}

final class AdditionalOrderRequest
{
    public function __construct(
        public readonly string $userId,
        public readonly string $amount,
    ) {}
}

final class OrderController
{
    use PresentationPreparation;

    public function __construct(
        private readonly NewOrderUsecase $newOrderUsecase,
        private readonly AdditionalBuyOrderUsecase $additionalBuyOrderUsecase,
    ) {}

    public function newOrder(NewOrderRequest $req): void
    {
        $uid = $this->parseUserId($req->userId);
        $amt = $this->parseAmount($req->amount);
        try {
            $this->newOrderUsecase->run(new NewOrderUsecaseInput($uid, $amt));
        } catch (NewOrderUserAlreadyExistsException $e) {
            throw new BadRequestException('user already has account');
        } catch (NewOrderAmountTooSmallException $e) {
            throw new BadRequestException('amount is too small');
        }
    }

    public function additionalOrder(AdditionalOrderRequest $req): void
    {
        $uid = $this->parseUserId($req->userId);
        $amt = $this->parseAmount($req->amount);
        try {
            $this->additionalBuyOrderUsecase->run(new AdditionalBuyOrderUsecaseInput($uid, $amt));
        } catch (AdditionalBuyOrderUserNotFoundException $e) {
            throw new BadRequestException('user has no live account');
        } catch (AdditionalBuyOrderAmountTooSmallException $e) {
            throw new BadRequestException('amount is too small');
        }
    }

}
